api/api-roles-config-v1/src/main/java/com/thoughtworks/go/apiv1/admin/security/representers/RoleRepresenter.java

```diff
@@ -16,87 +16,79 @@
-            modelAttributeType = Map.class,
-            modelAttributeName = "errors")
-    Map errors();
-
-
-    public class RoleTypeGetter implements BiFunction<Role, RequestContext, String> {
-        @Override
-        public String apply(Role role, RequestContext requestContext) {
-            if (role instanceof RoleConfig) {
-                return "gocd";
-            } else {
-                return "plugin";
-            }
-        }
+import java.util.*;
+
+import static com.thoughtworks.go.api.representers.RepresenterUtils.addLinks;
+
+
+public class RoleRepresenter {
+
+    private static List<Link> getLinks(Role model, RequestContext requestContext) {
+        return Arrays.asList(
+                new Link("doc", "https://api.gocd.org/#roles"),
+                requestContext.build("self", "/go/api/admin/security/roles/%s", model.getName()),
+                requestContext.build("find", "/go/api/admin/security/roles/:role_name")
+        );
     }
 
-    class RoleErrorGetter extends ErrorGetter {
-        public RoleErrorGetter() {
-            super(Collections.singletonMap("authConfigId", "auth_config_id"));
+    public static Map toJSON(Role role, RequestContext requestContext) {
+        if (role == null) return null;
+        Map<String, Object> jsonObject = new LinkedHashMap<>();
+        List<Link> links = getLinks(role, requestContext);
+        addLinks(links, jsonObject);
+
+        jsonObject.put("name", role.getName().toString());
+        jsonObject.put("type", getRoleType(role));
+        if (role.hasErrors()) {
+            jsonObject.put("errors", new ErrorGetter(Collections.singletonMap("authConfigId", "auth_config_id"))
+                    .apply(role, requestContext));
         }
+        if (role instanceof RoleConfig) {
+            jsonObject.put("attributes", GoCDRoleConfigRepresenter.toJSON((RoleConfig) role, requestContext));
+        } else if (role instanceof PluginRoleConfig) {
+            jsonObject.put("attributes", PluginRoleConfigRepresenter.toJSON((PluginRoleConfig) role, requestContext));
+        }
+        return jsonObject;
     }
 
-    class RoleConfigLinksProvider implements LinksProvider<Role> {
-        private static final Link DOC = new Link("doc", "https://api.gocd.org/#roles");
+    public static Role fromJSON(JsonObject jsonObject, RequestContext requestContext) {
+        Role model;
+        String type = jsonObject.get("type").getAsString();
 
-        @Override
-        public List<Link> getLinks(Role model, RequestContext requestContext) {
-            return Arrays.asList(
-                    DOC,
-                    requestContext.build("self", "/go/api/admin/security/roles/%s", model.getName()),
-                    requestContext.build("find", "/go/api/admin/security/roles/:role_name")
-            );
+        if ("gocd".equals(type)) {
+            model = GoCDRoleConfigRepresenter.fromJSON(jsonObject.get("attributes").getAsJsonObject());
```

Comment:
This seems to assume that `attributes` is always present in the JSON, even if the value is blank and is a potential for a NPE.